File: bench/micro.py
#!/usr/bin/env python3
"""Micro-benchmark: tokenize 'x gtb show         1257' with tiktoken."""
import os
import time
import tiktoken

TEXT = "x gtb show         1257"
ENCODING = "o200k_base"
RUNS = 10000

# Cold start: get_encoding + first encode in this process
t0 = time.perf_counter()
enc = tiktoken.get_encoding(ENCODING)
ids = enc.encode(TEXT)
t1 = time.perf_counter()
cold_us = (t1 - t0) * 1e6

# Hot loop
start = time.perf_counter()
for _ in range(RUNS):
    enc.encode(TEXT)
elapsed = time.perf_counter() - start

print("tiktoken Python")
print(f"  text:          {repr(TEXT)}")
print(f"  tokens:        {len(ids)}")
print(f"  token IDs:     {ids}")
print(f"  cold start:    {cold_us:.1f} µs")
print(f"  hot loop:      {elapsed*1e6/RUNS:.2f} µs/op  ({RUNS/elapsed:,.0f} ops/s)")

File: examples/micro.rs
use std::env;
use std::time::Instant;
use token_actuary::Actuary;

const TEXT: &str = "x gtb show         1257";
const RUNS: usize = 100_000;

fn main() {
    let tokenizer_path = env::var("TOKENIZER_JSON")
        .unwrap_or_else(|_| "../tokenizer-json/data/gpt_4o.tokenizer.json".into());

    // Cold: load + first encode
    let t0 = Instant::now();
    let actuary = Actuary::from_file(&tokenizer_path).unwrap();
    let ids = actuary.encode(TEXT, false).unwrap();
    let cold = t0.elapsed().as_micros();

    // Hot loop
    let t0 = Instant::now();
    for _ in 0..RUNS {
        let _ = actuary.encode(TEXT, false).unwrap();
    }
    let hot_total = t0.elapsed().as_micros() as f64;
    let hot_avg = hot_total / RUNS as f64;

    println!("ta Rust library");
    println!("  text:          {:?}", TEXT);
    println!("  tokens:        {}", ids.len());
    println!("  token IDs:     {:?}", ids);
    println!("  cold start:    {} µs", cold);
    println!("  hot loop:      {:.2} µs/op  ({:.0} ops/s)", hot_avg, 1e6 / hot_avg);
}
